Add option to ignore compared fields existence check

kssumin opened the pull request:
This adds a new option (ignoringNonExistentFields) to recursive comparison that allows polymorphic objects with different field structures to be compared effectively. When enabled, the comparison will skip field existence validation, which is useful when comparing objects of different subtypes where some fields might only exist in certain implementations.

## Overview
This PR adds a new feature to the recursive comparison functionality that allows ignoring non-existent field checks. This is particularly useful when comparing polymorphic objects where some fields might not exist in all subtypes.

## Implementation Details
- Added `ignoreNonExistentFields` boolean flag to `RecursiveComparisonConfiguration`
- Added `ignoringNonExistentFields()` method to `RecursiveComparisonAssert`
- Added appropriate builder methods and setters
- Added early return in `checkComparedFieldsExist` when this option is enabled
- Added corresponding toString/description methods to show when this option is active
- Added unit tests to verify the behavior

## Example Use Case
When comparing objects like:
```java
class BaseClass {
    String commonField = "common";
}

class SubClass1 extends BaseClass {
    String subField1 = "sub1";
}

class SubClass2 extends BaseClass {
    // 'subField1' doesn't exist here
    String subField2 = "sub2";
}
```

Regular recursive comparison would fail due to field existence checks. With this new option:
```java
SubClass1 actual = new SubClass1();
SubClass2 expected = new SubClass2();

// This will now pass
assertThat(actual)
    .usingRecursiveComparison()
    .ignoringNonExistentFields()
    .isEqualTo(expected);
```

#### Check List:
* Fixes #3806 
* Unit tests : YES
* Javadoc with a code example (on API only) : YES
* PR meets the [contributing guidelines](https://github.com/assertj/assertj/blob/main/CONTRIBUTING.md) : YES

joel-costigliola commented:
Thanks @kssumin, I will review your PR shortly !

joel-costigliola commented on assertj-core/src/test/java/org/assertj/core/internal/ConfigurableRecursiveFieldByFieldComparator_Test.java:
```diff
@@ -87,4 +87,48 @@ void should_throw_an_NPE_if_given_RecursiveComparisonConfiguration_is_null() {
   }
 
+  @Test
+  void should_success_ignore_non_existent_fields_when_configured() {
```
I'm not completely sure what we are testing here, we check that the config has the expected `ignoreNonExistentFields` value but I don't get the point of checking `int compare = configurableRecursiveFieldByFieldComparator.compare(actual, other);`.

Instead of these two tests, could we add a test in `RecursiveComparisonAssert_isEqualTo_comparingOnlyFields_Test` and reproduce the test example in the initial issue https://github.com/assertj/assertj/issues/3806#issuecomment-2745230319 ? if the implementation is correct, the test should pass

joel-costigliola replied:
We could in addition also add a test with the code example from the javadoc (with BaseClass, Subclass1 and 2)

Thanks !

joel-costigliola commented on assertj-core/src/main/java/org/assertj/core/api/RecursiveComparisonAssert.java:
```diff
@@ -649,6 +649,46 @@ public SELF ignoringActualEmptyOptionalFields() {
+   * @return this {@link RecursiveComparisonAssert} to allow fluent chaining.
+   */
+  @CheckReturnValue
+  public SELF ignoringNonExistentFields() {
```
I still need to think if we keep that name or if we should rename to `ignoringNonExistentComparedFields` which is more accurate.

joel-costigliola commented:
Thanks @kssumin, first round of code review done.

kssumin commented:
@joel-costigliola Thank you for your review and suggestions! I've made the following changes based on your feedback:

1. Added test cases in RecursiveComparisonAssert_isEqualTo_comparingOnlyFields_Test that reproduce the example from issue #3806
2. Added another test with the BaseClass/SubClass example from the JavaDoc to verify the functionality
3. Updated the method name as suggested to be more descriptive (ignoringNonExistentComparedFields)

I've maintained the test structure with GIVEN/WHEN/THEN pattern and made sure all tests verify both success and failure cases. Let me know if you'd like me to make any additional changes!

joel-costigliola commented:
Integrated thanks @kssumin !
